src/jit/hwintrinsicxarch.cpp

```diff
@@ -153,37 +160,139 @@ InstructionSet Compiler::isaOfHWIntrinsic(NamedIntrinsic intrinsic)
+    return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].ins[type - TYP_BYTE];
+}
+
+//------------------------------------------------------------------------
+// categoryOfHWIntrinsic: get the category of the given intrinsic
+//
+// Arguments:
+//    intrinsic -- id of the intrinsic function.
+//
+// Return Value:
+//     the category of the given intrinsic
+//
+HWIntrinsicCategory Compiler::categoryOfHWIntrinsic(NamedIntrinsic intrinsic)
+{
+    assert(intrinsic != NI_Illegal);
+    assert(intrinsic > NI_HW_INTRINSIC_START && intrinsic < NI_HW_INTRINSIC_END);
+    return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].category;
+}
+
+//------------------------------------------------------------------------
+// HWIntrinsicFlag: get the flag of the given intrinsic
+//
+// Arguments:
+//    intrinsic -- id of the intrinsic function.
+//
+// Return Value:
+//     the flag of the given intrinsic
+//
+HWIntrinsicFlag Compiler::flagOfHWIntrinsic(NamedIntrinsic intrinsic)
+{
+    assert(intrinsic != NI_Illegal);
+    assert(intrinsic > NI_HW_INTRINSIC_START && intrinsic < NI_HW_INTRINSIC_END);
+    return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].flag;
+}
+
+//------------------------------------------------------------------------
+// getArgForHWIntrinsic: get the argument from the stack and match  the signature
+//
+// Arguments:
+//    argType   -- the required type of argument
+//    argClass  -- the class handle of argType
+//
+// Return Value:
+//     get the argument at the given index from the stack and match  the signature
+//
+GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE argClass)
+{
+    GenTree* arg = nullptr;
+    if (argType == TYP_STRUCT)
     {
-        case NI_SSE_IsSupported:
-        case NI_SSE2_IsSupported:
-        case NI_SSE3_IsSupported:
-        case NI_SSSE3_IsSupported:
-        case NI_SSE41_IsSupported:
-        case NI_SSE42_IsSupported:
-        case NI_AVX_IsSupported:
-        case NI_AVX2_IsSupported:
-        case NI_AES_IsSupported:
-        case NI_BMI1_IsSupported:
-        case NI_BMI2_IsSupported:
-        case NI_FMA_IsSupported:
-        case NI_LZCNT_IsSupported:
-        case NI_PCLMULQDQ_IsSupported:
-        case NI_POPCNT_IsSupported:
-            return true;
-        default:
-            return false;
+        unsigned int argSizeBytes;
+        var_types    base = getBaseTypeAndSizeOfSIMDType(argClass, &argSizeBytes);
+        argType           = getSIMDTypeForSize(argSizeBytes);
+        assert(argType == TYP_SIMD32 || argType == TYP_SIMD16);
+        arg = impSIMDPopStack(argType);
+        assert(arg->TypeGet() == TYP_SIMD16 || arg->TypeGet() == TYP_SIMD32);
     }
+    else
+    {
+        assert(varTypeIsArithmetic(argType));
+        arg = impPopStack().val;
+        assert(varTypeIsArithmetic(arg->TypeGet()));
+        assert(genTypeSize(argType) <= genTypeSize(arg->TypeGet()));
```

Comment:
@CarolEidt @tannergooding Is this type validation enough? It seems hard to have a precise and general assert for all the HW intrinsics...

Reply:
I think we want something like `genActualType(arg->gtType) == genActualType(argType)`

Reply:
...at least for non-SIMD types. SIMD Types we probably want to be validating the baseType

Reply:
Yeah, this one looks better.

Reply:
Agree for the non-SIMD types. IIRC, we can't always rely on the stack node having its type normalized, which is why `impSIMDPopStack()` doesn't do that. I would recommend something like:
```
if (varTypeIsSIMD(argType))
{
    arg = impSIMDPopStack();
}
else
{
    arg = impPopStack().val;
    assert(genActualType(arg->gtType) == genActualType(argType));
}
```
If, in future, we change the normalizing of types to happen earlier, we can add that check in just `impSIMDPopStack()`.